Source/SwiftLintFramework/Extensions/Configuration+Remote.swift

```diff
@@ -0,0 +1,240 @@
+            throw ConfigurationError.generic("Invalid configuration entry: \"\(urlString)\" isn't a valid url.")
+        }
+
+        // Load from url
+        var taskResult: (Data?, URLResponse?, Error?)
+        var taskDone: Bool = false
+
+        // `.ephemeral` disables caching (which we don't want to be managed by the system)
+        let task = URLSession(configuration: .ephemeral).dataTask(with: url) { data, response, error in
+            taskResult = (data, response, error)
+            taskDone = true
+        }
+
+        task.resume()
+
+        let timeout = cachedFilePath == nil ? remoteConfigTimeout : remoteConfigTimeoutIfCached
+        let time = CFAbsoluteTimeGetCurrent()
+
+        // Block main thread until timeout is reached / task is done
+        while true {
+            if taskDone { break }
+            if CFAbsoluteTimeGetCurrent() - time > timeout { task.cancel(); break }
+            usleep(50_000) // Sleep for 50 ms
+        }
+
+        // Handle wrong data
+        guard
+            taskResult.2 == nil, // No error
+            (taskResult.1 as? HTTPURLResponse)?.statusCode == 200,
+            let configString = (taskResult.0.flatMap { String(data: $0, encoding: .utf8) })
+        else {
+            return try handleWrongData(
+                urlString: urlString,
+                cachedFilePath: cachedFilePath,
+                taskDone: taskDone,
+                timeout: timeout
+            )
+        }
+
+        // Handle file write failure
+        guard let filePath = cache(configString: configString, from: urlString, rootDirectory: rootDirectory) else {
+            return try handleFileWriteFailure(urlString: urlString, cachedFilePath: cachedFilePath)
+        }
+
+        // Handle success
+        self = .existing(path: filePath)
+        return filePath
+    }
+
+    private mutating func handleMissingNetwork(urlString: String, cachedFilePath: String?) throws -> String {
+        if let cachedFilePath = cachedFilePath {
+            queuedPrint(
+                "No internet connectivity: Unable to load remote config from \"\(urlString)\". "
+                    + "Using cached version as a fallback."
+            )
+            self = .existing(path: cachedFilePath)
+            return cachedFilePath
+        } else {
+            throw ConfigurationError.generic(
+                "No internet connectivity: Unable to load remote config from \"\(urlString)\". "
+                    + "Also didn't found cached version to fallback to."
+            )
+        }
+    }
+
+    private mutating func handleWrongData(
+        urlString: String,
+        cachedFilePath: String?,
+        taskDone: Bool,
+        timeout: TimeInterval
+    ) throws -> String {
+        if let cachedFilePath = cachedFilePath {
+            if taskDone {
+                queuedPrint("Unable to load remote config from \"\(urlString)\". Using cached version as a fallback.")
```

Comment:
@jpsim Should we drop a warning in cases where a new version from the remote could not be loaded or just print it?

I'm in favor of printing so that you're not spammed with warnings if running SwiftLint in an offline environment. In the end, that's the whole thing about caching, that it's there if needed...

Reply:
I'm leaning a bit more towards a warning than just a log. Otherwise the output of SwiftLint will be incorrect even for a valid configuration based on the host machine's conditions.